### PR TITLE
wq: openssl support

### PR DESCRIPTION
**Pull request by btovar**
(Issue #2635)

@dthain: If it looks ok, I can update the manual and man pages.

Still need to add the options to makeflow when using Work Queue, but I think we can do that in another pr.

I tried to make it so that current code does not need any change. Only a handful of functions from link.c had to be wrapped with openssl code.

For the catalog, the managers report "ssl: true when appropriate.  A worker can then initiate a TLS connection. When using HOST PORT, a worker needs the --ssl flag.  The --ssl flag is always needed for `work_queue_status HOST PORT` when the manager to query is using ssl.

### CODE REVIEW

**Comment by btovar** [comment]
Seems like I need to update the docker images to include libssl-dev

File: dttools/src/link.c [hunk 1]
```diff
@@ -346,13 +363,73 @@ void sockaddr_set_port( struct sockaddr_storage *addr, int port )
...
         }
-}               
+}
 
 struct link *link_serve(int port)
 {
 	return link_serve_address(0, port);
 }
 
+int _ssl_errors_cb(const char *str, size_t len, void *use_warn) {
```
Comment (dthain) [inline dttools/src/link.c:373]: static functions?

File: dttools/src/link.c [hunk 2]
```diff
@@ -560,13 +715,70 @@ struct link *link_connect(const char *addr, int port, time_t stoptime)
 	return 0;
 }
 
+ssize_t _read(struct link *link, char *data, size_t count) {
```
Comment (dthain) [inline dttools/src/link.c:718]: Name of _read likely conflicts with libc symbols.

File: dttools/src/link.c [hunk 3]
```diff
@@ -560,13 +715,70 @@ struct link *link_connect(const char *addr, int port, time_t stoptime)
...
+
+		return result;
+	}
+#endif
+
+	//if no ssl support, or no link->ssl, fallback to read syscall
+	return read(link->fd, data, count);
+}
+
+ssize_t _write(struct link *link, const char *data, size_t count) {
```
Comment (dthain) [inline dttools/src/link.c:745]: Ibid

**Review by dthain — CHANGES_REQUESTED** [review]
Yes, I think this is really well done, with the changes being quite compact.
A few naming items in the middle, but the approach looks good overall.

**Comment by btovar** [comment]
Build fails because TLS_client_method() in more recent versions of openssl. Adding a config check to select the correct openssl call...

**Comment by btovar** [comment]
@dthain, this is ready to merge from my side. You may want to take a look at the note in the wq manual. 
https://github.com/cooperative-computing-lab/cctools/pull/2733/commits/33db544c64b3f1231f5c754a926074eeba408353

File: doc/manuals/work_queue/index.md [hunk 4]
```diff
@@ -1222,6 +1222,56 @@ API](http://ccl.cse.nd.edu/software/manuals/api/html/work__queue_8h.html).
 
 ### Security
```
Comment (dthain) [inline doc/manuals/work_queue/index.md:1224]: At the beginning of the section, place the note that, by default, no authentication is done,
and so you can do SSL, password, or both.

File: doc/manuals/work_queue/index.md [hunk 5]
```diff
@@ -1222,6 +1222,56 @@ API](http://ccl.cse.nd.edu/software/manuals/api/html/work__queue_8h.html).
...
 
+#### TLS (SSL) support
+
+Work Queue can encrypt the communication between manager and workers using TLS.
+For this, you need to specify the key and certificate (in PEM format) of your
+server when creating the queue.
+
+If you do not have a key and certificate at hand, but you want the
+communications to be encrypted, you can create them using a command like:
+
```
Comment (dthain) [inline doc/manuals/work_queue/index.md:1233]: Explain a little bit more here that this is self-signed, with the usual proviso...

File: doc/manuals/work_queue/index.md [hunk 6]
```diff
@@ -1222,6 +1222,56 @@ API](http://ccl.cse.nd.edu/software/manuals/api/html/work__queue_8h.html).
...
+work_queue_worker (... other args ...) --ssl HOST PORT
+work_queue_factory (... other args ...) --ssl HOST PORT
+work_queue_status --ssl HOST PORT
+condor_submit_workers -E'--ssl' HOST PORT
+```
+
+
+#### Password Files
+
 By default, Work Queue does **not** perform any authentication, so any workers
```
Comment (dthain) [inline doc/manuals/work_queue/index.md:1275]: Move this paragraph above...

File: doc/manuals/work_queue/index.md (hunk 4 above)
Comment (dthain) [reply doc/manuals/work_queue/index.md:1224]: Also a bit that SSL provides manager-to-worker authentication, while password gives you worker-to-manager auth.

Comment (dthain) [reply doc/manuals/work_queue/index.md:1224]: Also, that reminds me that the warning message emitted by the manager should now mention using either password or ssl.

File: doc/manuals/work_queue/index.md [hunk 7]
```diff
@@ -1222,6 +1222,56 @@ API](http://ccl.cse.nd.edu/software/manuals/api/html/work__queue_8h.html).
 
 ### Security
 
+#### TLS (SSL) support
```
Comment (dthain) [inline doc/manuals/work_queue/index.md:1225]: Call this SSL consistently, since that's how we name the options and code...

**Pull request merged**